libraries/chain/controller.cpp

```diff
@@ -3999,18 +3985,138 @@ struct controller_impl {
+         EOS_ASSERT( f_ext.new_finalizer_policy_diff, invalid_qc_claim,
+                     "Savanna Genesis block #${b} finality block header extension misses new_finalizer_policy_diff",
+                     ("b", block_num) );
+
+         finalizer_policy no_policy;
+         // apply_diff will FC_ASSERT if new_finalizer_policy_diff is malformated
+         finalizer_policy genesis_policy = no_policy.apply_diff(*f_ext.new_finalizer_policy_diff);
+         EOS_ASSERT( genesis_policy.generation == 1, invalid_qc_claim,
+                     "Savanna Genesis block #${b} finalizer policy generation (${g}) not 1",
+                     ("b", block_num)("g", genesis_policy.generation) );
+      }
+   }
+
    // thread safe, expected to be called from thread other than the main thread
    template<typename ForkDB, typename BS>
    block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
-      constexpr bool savanna_mode = std::is_same_v<typename std::decay_t<BS>, block_state>;
-      if constexpr (savanna_mode) {
-         // Verify claim made by finality_extension in block header extension and
-         // quorum_certificate_extension in block extension are valid.
-         // This is the only place the evaluation is done.
-         verify_qc_claim(id, b, prev);
+      constexpr bool is_proper_savanna_block = std::is_same_v<typename std::decay_t<BS>, block_state>;
+      assert(is_proper_savanna_block == b->is_proper_svnn_block());
+
+      // If is_proper_savanna_block is true, then it means that the parent block of block b has a block_state.
+      //
+      // For a block to have a block_state it means the block must be a Savanna block,
+      // which means either Transition block or a Proper Savanna block.
+      //
+      // A block that has a Savanna block as a parent must also be a Savanna block.
+      // A block that has a Proper Savanna block as a parent must also be a Proper Savanna block.
+      //
+      // So given that the parent block has a block_state, we know that block b must either be
+      // a Transition block or a Proper Savanna block.
+
+      // If is_proper_savanna_block is false, then it means that the parent block of block b
+      // has a block_state_legacy.
+      //
+      // For a block to have a block_state_legacy it means the block must either be
+      // a Legacy block or a Transition block.
+      //
+      // A block that has a Legacy block as a parent must either be a Legacy block
+      // or a Transition block.
+      //
+      // A block that has a Transition block as a parent must either be a Transition
+      // block or a Proper Savanna block.
+
+      if constexpr (is_proper_savanna_block) {
+         EOS_ASSERT( b->is_proper_svnn_block(), block_validate_exception,
+                     "create_block_state_i cannot be called on block #${b} which is not a Savanna block while its parent is a Savanna block",
```

Comment:
Comment should be:

`"create_block_state_i cannot be called on block #${b} which is not a Proper Savanna block unless the prev block state provided is of type block_state"`